flow/ThreadHelper.actor.h

```diff
@@ -532,6 +532,66 @@ Future<T> unsafeThreadFutureToFuture(ThreadFuture<T> threadFuture) {
 	return callback->promise.getFuture();
 }
 
+// A callback waiting on a thread future and will delete itself once fired
+template <class T>
+struct UtilCallback : public ThreadCallback {
+public:
+	UtilCallback(ThreadFuture<T> f, void* userdata) : f(f), userdata(userdata) {}
+
+	bool canFire(int notMadeActive) const override { return true; }
+	void fire(const Void& unused, int& userParam) override {
+		g_network->onMainThread(Promise<Void>((SAV<Void>*)userdata), TaskPriority::DefaultOnMainThread);
+		delete this;
+	}
+	void error(const Error&, int& userParam) override {
+		g_network->onMainThread(Promise<Void>((SAV<Void>*)userdata), TaskPriority::DefaultOnMainThread);
+		delete this;
+	}
+	void destroy() override {}
+
+private:
+	ThreadFuture<T> f;
+	void* userdata;
+};
+
+// The underlying actor that converts ThreadFuture from Future
+// Note: should be used from main thread
+ACTOR template <class T>
+static Future<Void> safeThreadFutureToFutureActor(Promise<T> result, ThreadFuture<T> threadFuture) {
+	Promise<Void> ready;
+	Future<Void> onReady = ready.getFuture();
+	auto savPtr = ready.extractRawPointer();
+	UtilCallback<T>* callback = new UtilCallback<T>(threadFuture, savPtr);
+	int unused = 0;
+	threadFuture.callOrSetAsCallback(callback, unused, 0);
+	wait(onReady);
+	// threadFuture should be ready
+	if (threadFuture.isError())
+		result.sendError(threadFuture.getError());
+	result.send(threadFuture.get());
+	return Void();
+}
+
+// A wrapper actor used for cancellation
+ACTOR template <class T>
+static Future<T> safeThreadFutureToFutureCancellableActor(ThreadFuture<T> threadFuture) {
+	state Promise<T> result;
+	Future<Void> cancellable = safeThreadFutureToFutureActor(result, threadFuture);
+	threadFuture.getPtr()->setCancel(Future<Void>(cancellable));
+	wait(cancellable);
+	Future<T> ready = result.getFuture();
+	if (ready.isError())
```

Comment:
Has this been clang formatted? I don't actually know the formatting rules that well to know whether single line blocks with no braces are allowed, but I just wanted to double check.

Reply:
Yeah, it is formatted.
It seems to allow it. I can add the braces if it avoids the confusion.